kitodo/kitodo-production
# Performance MetaDataEditor Client-Side Selection State Without Reloading

Currently, when selecting an element in the meta data editor, the following components need to be updated via JSF in order to reflect this new selection state:
- the meta data panel
- the logical structure tree
- the physical structure tree
- the list of pages in the pagination panel
- the gallery stripes
- the vertical thumbnail list of the detail view

This pull request proposes to partially manage the selection state of elements on the client side (via Javascript). As a consequence, selecting an element will only require to reload the meta data of the selected element instead of a full reload of all components. This improves performance when selecting elements significantly.

For a process with 2000 images, the delay when clicking a thumbnail in the gallery is reduced from 3760ms to 820ms (desktop workstation) or from ~9000ms to 1560ms (laptop).

In order to achieve this, information about what DOM elements reference the same logical structure or image are needed. Therefore, the property `logicalTreeNodeId` is introduced, which provides this mapping between structure tree nodes and gallery thumbnails. When clicking a thumbnail in the gallery, the corresponding tree node of the logical structure tree is retrieved, and then dynamically styled by applying CSS classes to reflect this new selection. The same is done for all the other components (pagination list, physical structure tree, etc.).

### How to Test this Change

From a user's perspective there should be no difference to the previous selection behaviour. 

However, this pull request does not change the backend part of the selection management. Refreshing the page via F5 will reveal the server side selection state. If everything is correct, the selection should remain the same after refreshing the page. 

### Dependencies

This pull request depends on and includes the following two changes:
- #5143 
- #5146
